### ISSUE TITLE
`SELECT * FROM metrics` should return metric names, descriptions, etc.

### ISSUE THREAD

**Issue by betodealmeida** [DataJunction/datajunction#178]

**Comment by CircArgs** [comment]
@betodealmeida would you be able to add a bit of detail on this one? Even when selecting explicitly all metrics I get an error like `Metrics ... have non-shared parents` which I think makes sense to me but then this issue would not make sense

**Comment by betodealmeida** [comment]
Yeah, this one definitely needs more context. I added it here so I wouldn't forget.

The SQL API allows you to fetch data from metrics by running queries like:

```sql
SELECT
    "basic.num_comments"            -- a metric node
  , "basic.dimension.users.gender"  -- a column from a dimension node
FROM metrics                        -- a virtual table that encapsulates all metrics
WHERE
    "basic.dimension.users.country" != 'BR'
GROUP BY
     basic.dimension.users.gender
```

The query will be translated into SQL that actually be executed. It happens in [this function](https://github.com/DataJunction/dj/blob/8477abcee0ba5742dcb2a4b7b49e6171cf23ee4c/src/datajunction/sql/build.py#L221), and will produce SQL that looks like:

```sql
SELECT
    COUNT(*) AS "basic.num_comments"
  , gender AS "basic.dimension.users.gender"
FROM
    comments
JOIN
    users
ON
    comments.user_id = users.id
...
```

It would be nice to expand the abstraction of the `metrics` table, and allow queries like:

```sql
SELECT name, description FROM metrics
```

But I think it requires some discussion first of how we want those queries to look like. Additionally, it would be nice to improve the SQL API so that it can also run the SQL used to define nodes. For example, the node `basic.num_comments` has this SQL:

```sql
SELECT SUM(user_id) FROM basic.source.comments
```

Currently, we can't send that SQL to the SQL API, since today it can only be used to query the `metrics` virtual table. It would be nice to have that functionality, since eventually we could have a web UI where people can run the SQL for their nodes as they're developing them.

**Comment by CircArgs** [comment]
@betodealmeida thinking about `SELECT name, description FROM metrics` it seems like it could be confusing. Do you think there should be a differentiation like `metrics.metadata`? (credit to @samredai for this convention which makes a lot of sense to me). Either that or I suppose allow queries directly against the actual table so something like `select name, description from node where type='METRIC'`

**Comment by betodealmeida** [comment]
Yeah, I think there are 3 separate cases for the SQL API:

1. The original one, returning data for an arbitrary expression referencing metric nodes, that I implemented for [Apache Superset](https://superset.apache.org/).

```sql
SELECT "metric_node", "dimension_node.column"
FROM metrics
GROUP BY "dimension_node.column"
```

2. Returning metadata about nodes, metrics, databases, etc.

```sql
SELECT name, description FROM XXX
```

3. Evaluating arbitrary SQL from a node, to allow users to interactively test their nodes while authoring them:

```sql
SELECT COUNT(*) FROM some_node WHERE condition
```

i think the challenge here is making all these 3 uses cases unambiguous by introducing some kind of namespace for the `FROM` clause. Eg:

1. `SELECT metric_node FROM dj.metrics`
2. `SELECT name, description FROM dj.metadata`
3. `SELECT COUNT(*) FROM some_node`

For something like this we would need to prevent nodes from being created in a `dj` directory, to keep the cases distinct.

@agorajek @samredai thoughts here?

**Comment by samredai** [comment]
Agreed that we should think about how to make these different behaviors unambiguous because I think there are even more cases (or sub-cases?) to consider. For me, it's a bit hard to organize them all in my head. For example:

Joining nodes that share a common set of dimensions?
```sql
SELECT "metric_node_a", "metric_node_b", "dimension_node.common_column"
FROM metrics
GROUP BY "dimension_node.common_column"
```
Getting the history of a node that's been changed?
```sql
SELECT name, description, version FROM dj.metadata
```
Getting the raw SQL string from a transform node?
```sql
SELECT query FROM transform_node
```
I really like the idea of using namespacing to make this more intuitive and clear and @CircArgs also mentioned the idea of having a `dj` namespace which feels right and is probably the most flexible. Would everyone else find it helpful to have a docker compose setup that includes a superset container so we can play around with building visual components from these queries?

**Comment by betodealmeida** [comment]
> Joining nodes that share a common set of dimensions?
> 
> ```sql
> SELECT "metric_node_a", "metric_node_b", "dimension_node.common_column"
> FROM metrics
> GROUP BY "dimension_node.common_column"
> ```

This is possible today if `metric_node_a` and `metric_node_b` have the same parents.
 
> Getting the history of a node that's been changed?
> 
> ```sql
> SELECT name, description, version FROM dj.metadata
> ```

I like this one... we might want to introduce versioning early rather than later, since it's easier. This is another one of those cases where we might be over-engineering, but it's much easier to do it early.

> Getting the raw SQL string from a transform node?
> 
> ```sql
> SELECT query FROM transform_node
> ```

This one doesn't make sense from the current API perspective (or my proposal). If we're not using the `dj` namespace then the query is treated as a data query, so the query above would only make sense if `transform_node` has a column called `query`, and it would return its data.

We could do something similar, like:

```sql
SELECT query FROM dj.metadata WHERE name = 'transform_node'
``` 

> I really like the idea of using namespacing to make this more intuitive and clear and @CircArgs also mentioned the idea of having a `dj` namespace which feels right and is probably the most flexible. Would everyone else find it helpful to have a docker compose setup that includes a superset container so we can play around with building visual components from these queries?

Let me know if you need help connecting Superset to DJ!

**Comment by agorajek** [comment]
I think it may be easier to reason about all these metadata queries if we simply agree on the data model of these DJ metadata tables. I think making `dj` namespace restricted is a good start. Think of it like the catalog schema of any famous database, say Postgres, Oracle etc. 

Next let's have a look at some possible tables in there:
```sql
-- databases
dj.database (
  description STRING,
  uri URI,
  read_only BOOL,
  cost INT
)

-- all node
dj.nodes (
  description STRING,
  type STRING fk to dj.node_types.name, -- this maybe goes too far but to show the idea
  name STRING,
  query STRING,
  columns ARRAY<STRING>,
  attributes MAP<>,
  downstream_nodes ARRAY<STR>, -- list of all downstream nodes
  upstream_nodes ARRAY<STR>, -- list of all upstream nodes
)

-- dj.metric_nodes (similar to node, but specifically for metric type)
-- dj.transform_nodes (etc...)
-- dj.source_nodes (this one could be offering all the info we know about through the SQL interface, wow... this is so cool!)
```
With the above data model, there is no ambiguity what your queries should do:
1. If you query thing from the dj.* only, you should be getting results only from the DJ metadata space. I don't think we need an extra `dj.metadata` table btw, unless we come up with some special fun use for it. Example:
```sql
SELECT description, name FROM dj.nodes 
-- or from dj.metric_nodes, dj.transform_nodes, or dj.source_nodes
```
2. When you run query on the "other" objects in the DJ database, then you should be getting the user data, say:
```sql
SELECT * FROM basic.dimension.users 
-- should return all the data from that dimension
```
3. We could also offer some "hidden columns" on all the user nodes. Example:
```sql
SELECT __name__, __type__, __query__, user_id FROM basic.dimension.users 
-- which would repeat the same values on all the __*__ columns cause they are all the same
```

There is definitely also a place for combining queries on the dj metadata with the queries on the user data, but I don't have a good example in mind yet.
